### PR TITLE
CP-23901: Add CLI changes for VM.set_bios_strings

### PR DESCRIPTION
**Pull request by sharady**

### CODE REVIEW

**Comment by sharady** [comment]
It will fail for now, As it is based on https://github.com/xapi-project/xen-api/pull/3174
Will rebase it once above PR is merged.

File: ocaml/xapi/cli_operations.ml [hunk 1]
```diff
@@ -646,12 +646,13 @@ let make_param_funs getall getallrecs getbyuuid record class_name def_filters de
   let p_set (printer : printer) rpc session_id params =
     let record = get_record rpc session_id (List.assoc "uuid" params) in
     let set_params = List.filter (fun (p,_) -> not (List.mem p ("uuid"::stdparams))) params in
+    let set_map_params = Hashtbl.create 10 in
```
Comment (robhoes) [inline ocaml/xapi/cli_operations.ml]: Please add a comment that explains what exactly goes into this hash table. Also the name could be more descriptive.

A type annotation would help to understand this as well (`let ... : (..., ...) Hashtbl.t = ...`).

File: ocaml/xapi/cli_operations.ml [hunk 2]
```diff
@@ -662,11 +663,21 @@ let make_param_funs getall getallrecs getbyuuid record class_name def_filters de
...
-            if List.mem_assoc key map then remove_from_map key;
-            add_to_map key v
+            (* If set_map is present then accumulate all (key, value) pairs. *)
+            match field.set_map with
+            | Some set_map ->
+              if not (Hashtbl.mem set_map_params set_map) then
+                Hashtbl.add set_map_params set_map [(key, v)]
+              else begin
+                let existing_params = Hashtbl.find set_map_params set_map in
+                Hashtbl.replace set_map_params set_map (existing_params @ [(key, v)])
```
Comment (robhoes) [inline ocaml/xapi/cli_operations.ml]: `(key, v) :: existing_params` is a little better.

File: ocaml/xapi/cli_operations.ml [hunk 3]
```diff
@@ -662,11 +663,21 @@ let make_param_funs getall getallrecs getbyuuid record class_name def_filters de
...
-            let map = get_map () in
-            if List.mem_assoc key map then remove_from_map key;
-            add_to_map key v
+            (* If set_map is present then accumulate all (key, value) pairs. *)
+            match field.set_map with
+            | Some set_map ->
+              if not (Hashtbl.mem set_map_params set_map) then
+                Hashtbl.add set_map_params set_map [(key, v)]
+              else begin
+                let existing_params = Hashtbl.find set_map_params set_map in
```
Comment (robhoes) [inline ocaml/xapi/cli_operations.ml]: Slightly simpler would be to just have this else-clause, with `let existing_params = try Hashtbl.find ... with Not_found -> []`.

**Comment by sharady** [comment]
@robhoes thanks for the review, please have a look once more.

File: ocaml/xapi/cli_operations.ml [hunk 4]
```diff
@@ -646,27 +646,39 @@ let make_param_funs getall getallrecs getbyuuid record class_name def_filters de
...
-            let add_to_map = match field.add_to_map with Some f -> f | None -> failwith ("Map field '"^s^"' is read-only.") in
-            let remove_from_map = match field.remove_from_map with Some f -> f | None -> failwith (Printf.sprintf "Records broken (field %s)" s) in
-            let map = get_map () in
-            if List.mem_assoc key map then remove_from_map key;
-            add_to_map key v
+          (* If set_map is present then accumulate all (key, value) pairs into set_map_table *)
+          match field.set_in_map, field.set_map  with
+          | Some set_in_map, None -> set_in_map key v
+          | None, Some set_map ->
+            if not (Hashtbl.mem set_map_table set_map) then
```
Comment (robhoes) [inline ocaml/xapi/cli_operations.ml]: What I meant in my previous comment, was the this whole conditional block can be replaced by just this (`Hashtbl.replace` just adds the k/v pair if the key is not in the table):
```
let existing_params = try Hashtbl.find set_map_table set_map with Not_found -> [] in
Hashtbl.replace set_map_table set_map ((key, v) :: existing_params)
```

Comment (sharady) [reply ocaml/xapi/cli_operations.ml]: Updated, Thanks.

**Pull request merged**